File: Palace/OPDS/TPPOPDSIndirectAcquisition.m
#import "TPPOPDSIndirectAcquisition.h"

#import "TPPXML.h"
#import "Palace-Swift.h"

#pragma mark Dictionary Keys

static NSString *const NYPLOPDSIndirectAcquisitionTypeKey = @"type";

static NSString *const NYPLOPDSIndirectAcquisitionIndirectAcqusitionsKey = @"indirectAcquisitions";

#pragma mark -

@interface TPPOPDSIndirectAcquisition ()

@property (nonatomic, copy, nonnull) NSString *type;
@property (nonatomic, nonnull) NSArray<TPPOPDSIndirectAcquisition *> *indirectAcquisitions;

@end

@implementation TPPOPDSIndirectAcquisition

+ (instancetype _Nonnull)
indirectAcquisitionWithType:(NSString *const _Nonnull)type
indirectAcquisitions:(NSArray<TPPOPDSIndirectAcquisition *> *const _Nonnull)indirectAcquisitions
{
  return [[self alloc] initWithType:type indirectAcquisitions:indirectAcquisitions];
}

+ (instancetype _Nullable)indirectAcquisitionWithXML:(TPPXML *const _Nonnull)xml
{
  NSString *const type = [xml attributes][@"type"];
  if (!type) {
    return nil;
  }

  NSMutableArray<TPPOPDSIndirectAcquisition *> *const mutableIndirectAcquisitions = [NSMutableArray array];
  for (TPPXML *const indirectAcquisitionXML in [xml childrenWithName:@"indirectAcquisition"]) {
    TPPOPDSIndirectAcquisition *const indirectAcquisition =
      [TPPOPDSIndirectAcquisition indirectAcquisitionWithXML:indirectAcquisitionXML];

    if (indirectAcquisition) {
      [mutableIndirectAcquisitions addObject:indirectAcquisition];
    } else {
      TPPLOG(@"Ignoring invalid indirect acquisition.");
    }
  }

  return [self indirectAcquisitionWithType:type
                      indirectAcquisitions:[mutableIndirectAcquisitions copy]];
}

- (instancetype _Nonnull)initWithType:(NSString *const _Nonnull)type
                 indirectAcquisitions:(NSArray<TPPOPDSIndirectAcquisition *> *const _Nonnull)indirectAcquisitions
{
  self = [super init];

  self.type = type;
  self.indirectAcquisitions = indirectAcquisitions;

  return self;
}

+ (_Nullable instancetype)indirectAcquisitionWithDictionary:(NSDictionary *const _Nonnull)dictionary
{
  NSString *const type = dictionary[NYPLOPDSIndirectAcquisitionTypeKey];
  if (![type isKindOfClass:[NSString class]]) {
    return nil;
  }

  NSDictionary *const indirectAcquisitionDictionaries = dictionary[NYPLOPDSIndirectAcquisitionIndirectAcqusitionsKey];
  if (![indirectAcquisitionDictionaries isKindOfClass:[NSArray class]]) {
    return nil;
  }

  NSMutableArray *const mutableIndirectAcquisitions =
    [NSMutableArray arrayWithCapacity:indirectAcquisitionDictionaries.count];

  for (NSDictionary *const indirectAcquisitionDictionary in indirectAcquisitionDictionaries) {
    if (![indirectAcquisitionDictionary isKindOfClass:[NSDictionary class]]) {
      return nil;
    }

    TPPOPDSIndirectAcquisition *const indirectAcquisition =
    [TPPOPDSIndirectAcquisition indirectAcquisitionWithDictionary:indirectAcquisitionDictionary];
    if (!indirectAcquisition) {
      return nil;
    }

    [mutableIndirectAcquisitions addObject:indirectAcquisition];
  }

  return [self indirectAcquisitionWithType:type
                      indirectAcquisitions:[mutableIndirectAcquisitions copy]];
}

- (NSDictionary *_Nonnull)dictionaryRepresentation
{
  NSMutableArray *const mutableIndirectionAcqusitionDictionaries =
    [NSMutableArray arrayWithCapacity:self.indirectAcquisitions.count];

  for (TPPOPDSIndirectAcquisition *const indirectAcqusition in self.indirectAcquisitions) {
    [mutableIndirectionAcqusitionDictionaries addObject:[indirectAcqusition dictionaryRepresentation]];
  }

  return @{
    NYPLOPDSIndirectAcquisitionTypeKey: self.type,
    NYPLOPDSIndirectAcquisitionIndirectAcqusitionsKey: [mutableIndirectionAcqusitionDictionaries copy]
  };
}

@end

File: Palace/OPDS/TPPOPDSIndirectAcquisition.swift
import Foundation

@objc class TPPOPDSIndirectAcquisition: NSObject {

  @objc private(set) var type: String
  @objc private(set) var indirectAcquisitions: [TPPOPDSIndirectAcquisition]

  private static let typeKey = "type"
  private static let indirectAcquisitionsKey = "indirectAcquisitions"

  @objc init(type: String, indirectAcquisitions: [TPPOPDSIndirectAcquisition]) {
    self.type = type
    self.indirectAcquisitions = indirectAcquisitions
    super.init()
  }

  @objc static func indirectAcquisition(withType type: String, indirectAcquisitions: [TPPOPDSIndirectAcquisition]) -> TPPOPDSIndirectAcquisition {
    return TPPOPDSIndirectAcquisition(type: type, indirectAcquisitions: indirectAcquisitions)
  }

  @objc static func indirectAcquisition(withXML xml: TPPXML) -> TPPOPDSIndirectAcquisition? {
    guard let type = (xml.attributes as? [String: String])?["type"] else {
      return nil
    }

    var mutableIndirectAcquisitions = [TPPOPDSIndirectAcquisition]()
    for childXML in xml.childrenWithName("indirectAcquisition") {
      if let indirect = TPPOPDSIndirectAcquisition.indirectAcquisition(withXML: childXML) {
        mutableIndirectAcquisitions.append(indirect)
      } else {
        Log.log("Ignoring invalid indirect acquisition.")
      }
    }

    return indirectAcquisition(withType: type, indirectAcquisitions: mutableIndirectAcquisitions)
  }

  @objc static func indirectAcquisition(withDictionary dictionary: NSDictionary) -> TPPOPDSIndirectAcquisition? {
    guard let type = dictionary[typeKey] as? String else {
      return nil
    }

    guard let indirectAcquisitionDicts = dictionary[indirectAcquisitionsKey] as? [NSDictionary] else {
      return nil
    }

    var mutableIndirectAcquisitions = [TPPOPDSIndirectAcquisition]()
    for dict in indirectAcquisitionDicts {
      guard let indirect = TPPOPDSIndirectAcquisition.indirectAcquisition(withDictionary: dict) else {
        return nil
      }
      mutableIndirectAcquisitions.append(indirect)
    }

    return indirectAcquisition(withType: type, indirectAcquisitions: mutableIndirectAcquisitions)
  }

  @objc func dictionaryRepresentation() -> NSDictionary {
    let indirectDicts = indirectAcquisitions.map { $0.dictionaryRepresentation() }
    return [
      TPPOPDSIndirectAcquisition.typeKey: type,
      TPPOPDSIndirectAcquisition.indirectAcquisitionsKey: indirectDicts
    ] as NSDictionary
  }
}
